graphql/schemabuilder/pagination.go

```diff
@@ -389,95 +350,63 @@ func (sb *schemaBuilder) buildPaginatedField(typ reflect.Type, f interface{}) (*
-		return nil, fmt.Errorf("%s's args should be an object", funcType)
-	}
-
-	for name, nodeType := range inputObject.InputFields {
-		args[name] = nodeType
-	}
+	args, err := funcCtx.argsTypeMap(argType)
 
 	ret := &graphql.Field{
 		Resolve: func(ctx context.Context, source, args interface{}, selectionSet *graphql.SelectionSet) (interface{}, error) {
 
-			if _, ok := args.(ConnectionArgs); !ok {
+			val, ok := args.(ConnectionArgs)
+			if !ok {
 				return nil, fmt.Errorf("arguments should implement ConnectionArgs")
 			}
-
-			in := make([]reflect.Value, 0, funcType.NumIn())
-
-			if hasContext {
-				in = append(in, reflect.ValueOf(ctx))
-			}
-
-			// Set up source.
-			if hasSource {
-				sourceValue := reflect.ValueOf(source)
-				ptrSource := sourceValue.Kind() == reflect.Ptr
-				switch {
-				case ptrSource && !ptrFunc:
-					in = append(in, sourceValue.Elem())
-				case !ptrSource && ptrFunc:
-					copyPtr := reflect.New(typ)
-					copyPtr.Elem().Set(sourceValue)
-					in = append(in, copyPtr)
-				default:
-					in = append(in, sourceValue)
-				}
-			}
-
-			// Set up other arguments.
-			if val, _ := args.(ConnectionArgs); val.Args != nil {
-				in = append(in, reflect.ValueOf(val.Args).Elem())
+			funcCtx.hasArgs = val.Args != nil
+			var argsVal interface{}
+			if funcCtx.hasArgs {
+				argsVal = reflect.ValueOf(val.Args).Elem().Interface()
 			}
 
-			if hasSelectionSet {
-				in = append(in, reflect.ValueOf(selectionSet))
-			}
+			in := funcCtx.prepareResolveArgs(source, argsVal, ctx)
 
 			// Call the function.
 			out := fun.Call(in)
-			var result interface{}
-			if hasRet {
-				var err error
-				connectionArgs, _ := args.(ConnectionArgs)
-
-				result, err = getConnection(nodeKey, castSlice(out[0].Interface()), connectionArgs)
-				if err != nil {
-					return nil, err
-				}
-
-				out = out[1:]
-			} else {
-				result = true
-			}
-			if hasError {
-				if err := out[0]; !err.IsNil() {
-					return nil, err.Interface().(error)
-				}
-			}
 
-			if _, ok := retType.(*graphql.NonNull); ok {
-				resultValue := reflect.ValueOf(result)
-				if resultValue.Kind() == reflect.Ptr && resultValue.IsNil() {
-					return nil, fmt.Errorf("%s is marked non-nullable but returned a null value", funcType)
-				}
-			}
+			return funcCtx.extractPaginatedRetAndErr(nodeKey, out, args, retType)
 
-			return result, nil
 		},
 		Args:           args,
 		Type:           retType,
 		ParseArguments: argParser.Parse,
-		Expensive:      false,
+		Expensive:      funcCtx.hasContext,
 	}
 
 	return ret, nil
 }
 
+func (funcCtx *funcContext) extractPaginatedRetAndErr(nodeKey string, out []reflect.Value, args interface{}, retType graphql.Type) (interface{}, error) {
+
+	var result interface{}
+	connectionArgs, _ := args.(ConnectionArgs)
+
+	result, err := getConnection(nodeKey, castSlice(out[0].Interface()), connectionArgs)
+	if err != nil {
+		return nil, err
+	}
+	out = out[1:]
+	if funcCtx.hasError {
+		if err := out[0]; !err.IsNil() {
```

Comment:
interesting - what should we do if `getConnection` returns an error, but the resolver does not return an error?

i think we should still return an error, i.e. paginated fields should always possibly return errors.

Reply:
also, let's make sure to add tests for this case/ error cases in general.

Reply:
the paginated field is gonna return an error anyway this checks if the composed function had an error or not